realm/src/main/java/io/realm/base/BaseRealm.java

```diff
@@ -0,0 +1,493 @@
+        }
+        this.autoRefresh = autoRefresh;
+    }
+
+    /**
+     * Retrieve the auto-refresh status of the Realm instance.
+     * @return the auto-refresh status
+     */
+    public boolean isAutoRefresh() {
+        return autoRefresh;
+    }
+
+    /**
+     * Add a change listener to the Realm
+     *
+     * @param listener the change listener
+     * @see RealmChangeListener
+     */
+    public void addChangeListener(RealmChangeListener listener) {
+        checkIfValid();
+        for (WeakReference<RealmChangeListener> ref : changeListeners) {
+            if (ref.get() == listener) {
+                // It has already been added before
+                return;
+            }
+        }
+
+        changeListeners.add(new WeakReference<RealmChangeListener>(listener));
+    }
+
+    /**
+     * Remove the specified change listener
+     *
+     * @param listener the change listener to be removed
+     * @see io.realm.RealmChangeListener
+     */
+    public void removeChangeListener(RealmChangeListener listener) {
+        checkIfValid();
+        WeakReference<RealmChangeListener> weakRefToRemove = null;
+        for (WeakReference<RealmChangeListener> weakRef : changeListeners) {
+            if (listener == weakRef.get()) {
+                weakRefToRemove = weakRef;
+                // There won't be duplicated entries, checking is done when adding
+                break;
+            }
+        }
+        if (weakRefToRemove != null) {
+            changeListeners.remove(weakRefToRemove);
+        }
+    }
+
+    /**
+     * Remove all user-defined change listeners
+     *
+     * @see io.realm.RealmChangeListener
+     */
+    public void removeAllChangeListeners() {
+        checkIfValid();
+        changeListeners.clear();
+    }
+
+    protected void removeHandler(Handler handler) {
+        handler.removeCallbacksAndMessages(null);
+        handlers.remove(handler);
+    }
+
+    private void sendNotifications() {
+        Iterator<WeakReference<RealmChangeListener>> iterator = changeListeners.iterator();
+        List<WeakReference<RealmChangeListener>> toRemoveList = null;
+        while (iterator.hasNext()) {
+            WeakReference<RealmChangeListener> weakRef = iterator.next();
+            RealmChangeListener listener = weakRef.get();
+            if (listener == null) {
+                if (toRemoveList == null) {
+                    toRemoveList = new ArrayList<WeakReference<RealmChangeListener>>(changeListeners.size());
+                }
+                toRemoveList.add(weakRef);
+            } else {
+                listener.onChange();
+            }
+        }
+        if (toRemoveList != null) {
+            changeListeners.removeAll(toRemoveList);
+        }
+    }
+    // Reference to a file has been released.
+    protected static boolean isFileOpen(RealmConfiguration configuration) {
+        Integer refCount = globalRealmFileReferenceCounter.get(configuration.getPath());
```

Comment:
Good observations. 
This method only reads from `globalRealmFileReferenceCounter`,  so it should be safe?
Also both `compactRealm` and `deleteRealm` will only be available from Realm, not dynamic Realm, so just locking on Realm should be safe enough. It won't prevent from misuse of the API though, but internal data structures should be safe enough.

That said, we currently have a high number of internal caches/counters that make reasoning about thread safety really tricky. Maybe we should consider hiding the complexity behind some RealmCacheController or something?